kernel/src/grant.rs

```diff
@@ -343,7 +361,94 @@ impl KernelManagedLayout {
         // The location of the grant data T is the last element in the entire
         // grant region. Caller must verify that memory is accessible and well
         // aligned to T.
-        NonNull::new_unchecked(base_ptr.add(grant_size - grant_t_size.0))
+        let grant_t_size_usize: usize = grant_t_size.0;
+        NonNull::new_unchecked(base_ptr.add(grant_size - grant_t_size_usize))
+    }
+
+    /// Read an 8 bit value from the counter field offset by the specified
+    /// number of bits. This is a helper function for reading the counter field.
+    fn get_counter_offset(&self, offset_bits: usize) -> usize {
+        // # Safety
+        //
+        // Creating a `KernelManagedLayout` object requires that the pointers
+        // are well aligned and point to valid memory.
+        let counters_val = unsafe { self.counters_ptr.read() };
+        (counters_val >> offset_bits) & 0xFF
+    }
+
+    /// Return the number of upcalls stored by the core kernel for this grant.
+    fn get_upcalls_number(&self) -> usize {
+        self.get_counter_offset(0)
+    }
+
+    /// Return the number of read-only allow buffers stored by the core kernel
+    /// for this grant.
+    fn get_allow_ro_number(&self) -> usize {
+        self.get_counter_offset(8)
+    }
+
+    /// Return the number of read-write allow buffers stored by the core kernel
+    /// for this grant.
+    fn get_allow_rw_number(&self) -> usize {
+        self.get_counter_offset(16)
+    }
+
+    /// Return mutable access to the slice of stored upcalls for this grant.
+    /// This is necessary for storing a new upcall.
+    fn get_upcalls_slice(&mut self) -> &mut [SavedUpcall] {
+        // # Safety
+        //
+        // Creating a `KernelManagedLayout` object ensures that the pointer to
+        // the upcall array is valid.
+        unsafe { slice::from_raw_parts_mut(self.upcalls_array, self.get_upcalls_number()) }
+    }
```

Comment:
Calling this function twice in the same scope would give the caller multiple mutable references to the same memory, which is UB, without the caller having to use `unsafe`. This function should not be marked safe, and should document that callers are responsible for ensuring that only one reference to the `SavedUpcall` memory may exist at any time. Alternatively, we could get rid of this helper function, since it is only called once.

The same comment applies to the three functions below this as well.

Reply:
I added this helper function to better encapsulate the unsafe pointer portions, so they are not as mingled with all of the rest of the somewhat complex grant code. But what you bring up is a concern with this approach.

My attempt to address this is to require `&mut self` for this function, and then note in the `KernelManagedLayout` constructors that part of the safety contract there is that only one can be created per `base_ptr`. Does that logic work?

We might still want to change this even if my thinking is ok.

Reply:
I think that logic would work if we only needed to create one `KernelManagedLayout` per base pointer, but currently it looks like we create one each time `enter_grant_kernel_managed` is called, which is called in `subscribe()`, `allow_ro()` and `allow_rw()` -- it seems like for a given base_ptr each of these might be called once or multiple times.

I think what we actually want to say is "only one KernelManagedLayout per base_ptr can exist at any time", but it does seem like that requirement might force us to propagate `unsafe` to several of the places where we currently create those layouts.

The easiest solution might actually be to just not create the slice references at all and instead encapsulate those accesses in some safe function that just uses raw pointer reads, but it does seem like a shame to have to do so.